In [450]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [451]:
rand_arr = np.random.randint(0,10, 10)
rand_arr

array([9, 4, 7, 0, 7, 8, 2, 4, 2, 1])

In [452]:
np.where((rand_arr > 3) & (rand_arr < 8), rand_arr*(-1), rand_arr)

array([ 9, -4, -7,  0, -7,  8,  2, -4,  2,  1])

2. Заменить максимальный элемент случайного массива на 0

In [453]:
np.where(rand_arr == np.amax(rand_arr), 0, rand_arr)

array([0, 4, 7, 0, 7, 8, 2, 4, 2, 1])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [454]:
A = np.random.randint(0, 10, (3,3))
A

array([[0, 9, 5],
       [5, 4, 4],
       [5, 7, 0]])

In [455]:
B = np.random.randint(0, 10, (3,3))
B

array([[7, 0, 0],
       [6, 4, 4],
       [1, 0, 1]])

In [456]:
np.multiply(A, B)

array([[ 0,  0,  0],
       [30, 16, 16],
       [ 5,  0,  0]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [457]:
A = np.random.randint(0, 10, (8,3))
A

array([[9, 5, 1],
       [6, 1, 0],
       [0, 2, 9],
       [0, 7, 4],
       [0, 2, 0],
       [5, 4, 7],
       [3, 2, 3],
       [4, 5, 6]])

In [458]:
B = np.random.randint(0, 10, (2,2))
B

array([[5, 5],
       [9, 4]])

In [459]:
L1 = np.where(np.reshape(np.in1d(A, B[0]), (8,3)).sum(axis=1), True, False)
L2 = np.where(np.reshape(np.in1d(A, B[1]), (8,3)).sum(axis=1), True, False)
rows = (L1 * L2 == True).nonzero()[0]
rows

array([0, 5, 7])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [460]:
A = np.random.randint(0, 3, (10,3))
A

array([[1, 2, 0],
       [0, 0, 2],
       [2, 0, 0],
       [2, 2, 0],
       [0, 1, 1],
       [0, 1, 0],
       [2, 2, 0],
       [2, 1, 2],
       [1, 2, 2],
       [2, 2, 1]])

In [461]:
L = np.where(
    (
    (A[:,0] != A[:,1]) |
    (A[:,1] != A[:,2]) |
    (A[:,0] != A[:,2])),
    True, False
).nonzero()[0]
print(L)
A[L]

[0 1 2 3 4 5 6 7 8 9]


array([[1, 2, 0],
       [0, 0, 2],
       [2, 0, 0],
       [2, 2, 0],
       [0, 1, 1],
       [0, 1, 0],
       [2, 2, 0],
       [2, 1, 2],
       [1, 2, 2],
       [2, 2, 1]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [462]:
A = np.random.randint(0, 3, (10,2))
A

array([[2, 0],
       [1, 0],
       [1, 1],
       [1, 2],
       [2, 2],
       [2, 1],
       [2, 2],
       [1, 2],
       [1, 1],
       [2, 1]])

______
______

In [463]:
np.unique(A, axis=0)

array([[1, 0],
       [1, 1],
       [1, 2],
       [2, 0],
       [2, 1],
       [2, 2]])

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

**With Numpy**

In [464]:
A = np.random.randint(0, 10, (3,4))
A

array([[4, 7, 7, 0],
       [7, 8, 4, 7],
       [9, 3, 2, 8]])

In [465]:
np.prod(np.diag(A)[np.diag(A).nonzero()[0]])

64

**Without Numpy**

In [466]:
import random, math
A = [[random.randint(0, 9) for j in range(3)] for i in range(4)]
A

[[1, 9, 3], [1, 3, 1], [6, 5, 8], [2, 7, 0]]

In [467]:
diag = [A[i][i] for i in range(len(A[0])) if A[i][i] != 0]
math.prod(diag)

24

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

**With Numpy**

In [468]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
np.all(np.in1d(x, y)*np.in1d(y, x))

True

**Without Numpy**

In [469]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
var_bool = True
for i in x:
    for a in y:
        if not i in y or not a in x:
            var_bool = False
            break
var_bool

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

**With Numpy**

In [470]:
A = np.random.randint(0, 10, 20)
A

array([1, 3, 0, 0, 9, 1, 7, 9, 4, 9, 2, 3, 6, 4, 0, 8, 7, 1, 7, 4])

In [471]:
np.max(A[(A[0:-1] == 0).nonzero()[0]+1, np.newaxis]) # A[0:-1] - обрезаем последний элемент, чтобы не выйти за пределы массива, если он будет равен нулю

9

In [472]:
A = [random.randint(0, 9) for i in range(20)]

A

[0, 3, 4, 4, 3, 6, 7, 8, 2, 3, 1, 3, 5, 5, 4, 7, 2, 7, 0, 3]

**Without Numpy**

In [473]:
list_num = [A[i+1] for i in range(len(A)) if A[i] == 0 and i < len(A)-1]
if list_num:
    print(max(list_num))
else:
    print("Empty list")

3


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

**With Numpy**

In [474]:
A =  np.array([2, 2, 2, 3, 3, 3, 5]) # np.random.randint(0, 3, 20)
print(A)

[2 2 2 3 3 3 5]


In [475]:
np.unique(A, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

**Without Numpy**

In [476]:
A = [2, 2, 2, 3, 3, 3, 5]
list_uni = list(set(A))
list_count = [A.count(i) for i in list_uni]
(list_uni, list_count)

([2, 3, 5], [3, 3, 1])

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

**With Numpy**

In [477]:
X = np.random.randint(0, 10, (3, 3))
Y = np.random.randint(0, 10, (3, 3))
#print(X, "\n\n\n", Y)

In [478]:
P = np.add.outer(np.sum(X**2, axis=1), np.sum(Y**2, axis=1))
N = np.dot(X, Y.T)
np.sqrt(P - 2*N)

array([[ 3.60555128, 12.08304597, 12.08304597],
       [ 7.28010989,  9.16515139,  8.1240384 ],
       [ 9.2736185 ,  8.06225775,  9.21954446]])

**Without Numpy**

In [479]:
import scipy
scipy.spatial.distance.cdist(X, Y) #быстрее

array([[ 3.60555128, 12.08304597, 12.08304597],
       [ 7.28010989,  9.16515139,  8.1240384 ],
       [ 9.2736185 ,  8.06225775,  9.21954446]])

Использование scipy оказалось быстрее

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [480]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [481]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [482]:
mean_calories= calorie_stats.mean()
mean_calories

106.88311688311688

In [483]:
average_calories = mean_calories/60
average_calories # во столько раз больше у конкурентов

1.7813852813852813

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [484]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [485]:
median_calories = np.median(calorie_stats)
print(f"median = {median_calories};\nmean = {mean_calories};")

median = 110.0;
mean = 106.88311688311688;


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [486]:
all_percentile = np.percentile(calorie_stats,  range(100))
all_percentile

array([ 50. ,  50. ,  50. ,  55.6,  70. ,  70. ,  75.6,  83.2,  90. ,
        90. ,  90. ,  90. ,  90. ,  90. ,  90. ,  90. ,  91.6,  99.2,
       100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. ,
       100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. ,
       100. , 100. , 100. , 106.4, 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 115.2, 120. , 120. , 120. ,
       120. , 120. , 120. , 120. , 120. , 120. , 120. , 120. , 120. ,
       124. , 130. , 130. , 136.8, 140. , 140. , 140. , 147.2, 150. ,
       152.4])

In [487]:
nth_percentile = min(all_percentile[all_percentile > 60])

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [488]:
more_60_callories = np.count_nonzero(calorie_stats[calorie_stats > 60])
all_callories = np.count_nonzero(calorie_stats)
more_calories = more_60_callories/all_callories*100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [489]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

Среднее количество калорий отклоняется от среднего значения на calorie_std, наше же значение лежит вне промежутка mean +- std, а это значит, что мы довольно уникальная компания на рынке()

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Мы можем продвигать наш продукт как диетический, тк по собранным нами данным, можно прийти к выводу, что CrunchieMunchies один из лучших на рынке по количеству калорий на одну порцию.
Наш продукт в 1.78 более диетический чем средняя порция хлопьев, 96.1% компаний не получило значение калорийности как у нас, так же можем показать значения mean_calories, median_calories и calorie_std

Как-то так